vm/src/stdlib/sre/interp.rs

```diff
@@ -0,0 +1,1494 @@
+        let marks_index = 2 * group_index;
+        if marks_index + 1 < self.marks.len() {
+            (self.marks[marks_index], self.marks[marks_index + 1])
+        } else {
+            (None, None)
+        }
+    }
+    fn marks_push(&mut self) {
+        self.marks_stack.push((self.marks.clone(), self.lastindex));
+    }
+    fn marks_pop(&mut self) {
+        let (marks, lastindex) = self.marks_stack.pop().unwrap();
+        self.marks = marks;
+        self.lastindex = lastindex;
+    }
+    fn marks_pop_keep(&mut self) {
+        let (marks, lastindex) = self.marks_stack.last().unwrap().clone();
+        self.marks = marks;
+        self.lastindex = lastindex;
+    }
+    fn marks_pop_discard(&mut self) {
+        self.marks_stack.pop();
+    }
+
+    pub fn pymatch(mut self) -> Self {
+        let ctx = MatchContext {
+            string_position: self.start,
+            string_offset: self.string.offset(0, self.start),
+            code_position: 0,
+            has_matched: None,
+            toplevel: true,
+        };
+        self.context_stack.push(ctx);
+
+        let mut dispatcher = OpcodeDispatcher::new();
+        let mut has_matched = None;
+
+        loop {
+            if self.context_stack.is_empty() {
+                break;
+            }
+            let ctx_id = self.context_stack.len() - 1;
+            let mut drive = StackDrive::drive(ctx_id, self);
+
+            has_matched = dispatcher.pymatch(&mut drive);
+            self = drive.take();
+            if has_matched.is_some() {
+                self.popped_context = self.context_stack.pop();
+            }
+        }
+
+        self.has_matched = has_matched;
+        self
+    }
+
+    pub fn search(mut self) -> Self {
+        // TODO: optimize by op info and skip prefix
+        while self.start <= self.end {
+            self.match_all = false;
+            self = self.pymatch();
+
+            if self.has_matched == Some(true) {
+                return self;
+            }
+            self.start += 1;
+            self.reset();
+        }
+
+        self
+    }
+}
+
+#[derive(Debug, Clone, Copy)]
+pub(crate) enum StrDrive<'a> {
+    Str(&'a str),
+    Bytes(&'a [u8]),
+}
+impl<'a> StrDrive<'a> {
+    fn offset(&self, offset: usize, skip: usize) -> usize {
+        match *self {
+            StrDrive::Str(s) => s
+                .get(offset..)
+                .and_then(|s| s.char_indices().nth(skip).map(|x| x.0 + offset))
+                .unwrap_or_else(|| s.len()),
+            StrDrive::Bytes(b) => std::cmp::min(offset + skip, b.len()),
+        }
+    }
+
+    pub fn count(&self) -> usize {
+        match *self {
+            StrDrive::Str(s) => s.chars().count(),
+            StrDrive::Bytes(b) => b.len(),
+        }
+    }
+
+    fn peek(&self, offset: usize) -> u32 {
+        match *self {
+            StrDrive::Str(s) => unsafe { s.get_unchecked(offset..) }.chars().next().unwrap() as u32,
```

Comment:
Is this `unsafe` necessary? `chars()` expects not to be in the middle of a utf8 codepoint, so this could break in some weird ways if it's not correct.